src/QsCompiler/Transformations/FunctorGeneration.cs

```diff
@@ -148,6 +152,125 @@ public override QsStatementKind OnConjugation(QsConjugation stm)
         }
     }
 
+    /// <summary>
+    /// Scope transformation that splits any nested operation calls into separate statements so
+    /// that they can be properly reversed. This is necessary to avoid out of order execution of the
+    /// automatically generated adjoint. It is safe to do because an adjointable operation must return
+    /// Unit, so any nested calls can be replaced by Unit and those calls moved to separate,
+    /// ordered statements.
+    /// </summary>
+    public class ExtractNestedOperationCalls
+    : SyntaxTreeTransformation<ExtractNestedOperationCalls.TransformationsState>
+    {
+        public class TransformationsState
+        {
+            /// <summary>
+            /// Accumulates statements that have been lifted from the current statement.
+            /// </summary>
+            public List<QsStatement> AdditionalStatements = new List<QsStatement>();
+
+            public Stack<TypedExpression> CurrentExpression = new Stack<TypedExpression>();
+        }
+
+        public ExtractNestedOperationCalls()
+        : base(new TransformationsState())
+        {
+            this.Statements = new ExtractNestedOperationCalls.StatementTransformation(this);
+            this.Expressions = new ExtractNestedOperationCalls.ExpressionTransformation(this);
+            this.ExpressionKinds = new ExtractNestedOperationCalls.ExpressionKindTransformation(this);
+        }
+
+        public class StatementTransformation
+        : StatementTransformation<TransformationsState>
+        {
+            public StatementTransformation(SyntaxTreeTransformation<TransformationsState> parent)
```

Comment:
Do you need this constructor? I suppose if the class is public, giving people the option is nice, but if you make it private as per my other suggestion, then I think this constructor may go unused.
This goes for the expression kind and expression transformations class constructors as well.

Reply:
I think you're right... let me try taking those out.

Reply:
Actually, for some reason it's the other constructor I can remove and this one needs to be kept. It will compile but fail the tests, even though I update the calls above to use the remaining constructor.

Reply:
Oh whoops. You're right, I commented the wrong one of the two types of constructors. You have removed the one I meant.